### PR TITLE
Calculator mode (RFC)

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
A simple SELECT query can be written with implicit SELECT to enable calculator-style expressions, e.g., `ch "1 + 2"`. This is controlled by a new setting, `implicit_select`.

Potential downsides:
- it currently does not work with the FORMAT clause;
- if SELECT is written with a typo, e.g., `SELEC`, the error message is worse.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 350208b668241c6514de016010b1ce539ba4b198 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68502/350208b668241c6514de016010b1ce539ba4b198/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/67bbe6fe98c41958067bf68370232a0eb24af1f7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Server/PostgreSQLHandler.cpp [hunk 1]
```diff
@@ -286,7 +286,8 @@ void PostgreSQLHandler::processQuery()
             settings.max_query_size,
             settings.max_parser_depth,
             settings.max_parser_backtracks,
-            settings.allow_settings_after_format_in_insert);
+            settings.allow_settings_after_format_in_insert,
+            settings.implicit_select);
```
Comment (novikd) [inline src/Server/PostgreSQLHandler.cpp:290]: Why would somebody use it with PostgreSQL protocol? Maybe pass `false` here?

File: src/Parsers/ParserSelectQuery.cpp [hunk 2]
```diff
@@ -125,7 +125,11 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     {
         bool has_all = false;
         if (!s_select.ignore(pos, expected))
-            return false;
+        {
+            /// This allows queries without SELECT, like `1 + 2`.
+            if (!implicit_select || with_expression_list || tables)
```
Comment (novikd) [inline src/Parsers/ParserSelectQuery.cpp:130]: I don't understand why the condition contains `with_expression_list || tables`

**Review by novikd — COMMENTED** [review]
Also, it requires adding documentation. Otherwise, the feature will be lost.

File: src/Core/Settings.h [hunk 3]
```diff
@@ -900,6 +900,7 @@ class IColumn;
     M(Bool, restore_replace_external_engines_to_null, false, "Replace all the external table engines to Null on restore. Useful for testing purposes", 0) \
     M(Bool, restore_replace_external_table_functions_to_null, false, "Replace all table functions to Null on restore. Useful for testing purposes", 0) \
     M(Bool, create_if_not_exists, false, "Enable IF NOT EXISTS for CREATE statements by default", 0) \
+    M(Bool, implicit_select, false, "Allow writing simple SELECT queries without the leading SELECT keyword, which makes it simple for calculator-style usage, e.g. `1 + 2` becomes a valid query.", 0) \
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Core/Settings.h:903]: The client is already a REPL, so naming the new behavior as "repl_mode" would be confusing. While implicit SELECT is exactly what it does.

File: src/Server/PostgreSQLHandler.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply src/Server/PostgreSQLHandler.cpp:290]: Currently, the PostgreSQL protocol serves as an interface that accepts the same queries as the ClickHouse protocol, using the same syntax. That's why we apply the same rule here.

File: src/Parsers/ParserSelectQuery.cpp (hunk 2 above)
Comment (alexey-milovidov) [reply src/Parsers/ParserSelectQuery.cpp:130]: With WITH statement, queries without SELECT became ugly and ambiguous, that's why we want to forbid them.

If a query has FROM, it's better to be conservative for now and still require SELECT.

**Comment by alexey-milovidov** [comment]
I plan to merge this without enabling it by default but then use it as a building block together with other features, such as https://github.com/ClickHouse/ClickHouse/issues/65023 to enable queries like:

```
cat data.json | ch 'json.x.y[1]'
```

**Comment by alexey-milovidov** [comment]
Needs #70289.

**Pull request merged**